overlord/assertstate/assertstate.go

```diff
@@ -889,6 +892,68 @@ func TryEnforceValidationSets(st *state.State, validationSets []string, userID i
 	return addCurrentTrackingToValidationSetsHistory(st)
 }
 
+// EnforceLocalValidationSets checks the specified and currently tracked
+// validation sets against installed snaps and enforces them if the check is
+// successful. The validation sets aren't fetched so they must be in the DB.
+func EnforceLocalValidationSets(st *state.State, validationSets []string, snaps []*snapasserts.InstalledSnap, ignoreValidation map[string]bool) error {
```

Comment:
local is a bit of a strange term because we use in other places to mean validation sets that were acked locally and are not in the store

Reply:
also will this work actually ? are we missing an integration test?  I am a bit confused TryEnforceValidationSets will not have stored things in the database if checkBeforeCommit failed which is our scenario? we would kind of need to capture the things that were in the Backstore passed to it, no? not sure what would best there. We need to capture more state in some form about what we tried to do and need to adjust to

Reply:
Yes, I think you're right. I think I was misunderstanding (am probably still am) how these assertions and handled. If the validation sets in the conflict error returned by `TryEnforce` have specific sequences, we could use those when doing the enforcement. AFAICT, unlike the validation sets coming from the API, the stored assertions always have sequences (but please correct me if I'm wrong). So we can enforce the specific validation sets that were returned in the conflict error so there's no chance of a race condition invalidating the snaps we installed/refreshed